webapp/src/js/services/telemetry.js

```diff
@@ -0,0 +1,215 @@
+var moment = require('moment'),
+  uuid = require('uuid/v4');
+
+angular
+  .module('inboxServices')
+  .factory('Telemetry', function($log, $q, $window, DB, Session) {
+    'use strict';
+    'ngInject';
+
+    var queue = $q.resolve();
+
+    // Intentionally scoped to the whole browser (for this domain). We can then
+    // tell if multiple users use the same device
+    var DEVICE_ID_KEY = 'medic-telemetry-device-id';
+
+    // Intentionally scoped to the specific user, as they may perform a
+    // different role (online vs. offline being being the most obvious) with
+    // different performance implications
+    var DB_ID_KEY = ['medic', Session.userCtx().name, 'telemetry-db'].join('-');
+    var LAST_AGGREGATED_DATE_KEY = [
+      'medic',
+      Session.userCtx().name,
+      'telemetry-date',
+    ].join('-');
+
+    var getDb = function() {
+      var dbName = $window.localStorage.getItem(DB_ID_KEY);
+      if (!dbName) {
+        dbName =
+          'medic-user-' + Session.userCtx().name + '-telemetry-' + uuid();
+        $window.localStorage.setItem(DB_ID_KEY, dbName);
+      }
+      return $window.PouchDB(dbName); // avoid angular-pouch as digest isn't necessary here
+    };
+
+    var getUniqueDeviceId = function() {
+      var uniqueDeviceId = $window.localStorage.getItem(DEVICE_ID_KEY);
+      if (!uniqueDeviceId) {
+        uniqueDeviceId = uuid();
+        $window.localStorage.setItem(DEVICE_ID_KEY, uniqueDeviceId);
+      }
+
+      return uniqueDeviceId;
+    };
+
+    var getLastAggregatedDate = function() {
+      var date = parseInt(
+        $window.localStorage.getItem(LAST_AGGREGATED_DATE_KEY)
+      );
+      if (!date) {
+        date = Date.now();
+        $window.localStorage.setItem(LAST_AGGREGATED_DATE_KEY, date);
+      }
```

Comment:
It makes sense to me that `DEVICE_ID_KEY` might be lazily initialised in the getter, but less so for the last aggregated date.  Also I can't see any other place where `LAST_AGGREGATED_DATE_KEY` gets set, so this would have the value of the first call to `getLastAggregatedDate()` (although I haven't quite reasoned what this would be).

I think this should probably be set in the `aggregate()` function, and _only there_.

Reply:
Initialising them in `aggregate()` as Alex suggested makes sense to me. I think I liked that the `LAST_AGGREGATED_DATE_KEY` was only ever used in this one function, but the name of the function doesn't make it clear that it also sets the value so it's a bit unexpected.